src/components/chat/answer/feedbackButtons.tsx

```diff
@@ -6,35 +7,60 @@ import {
   faThumbsDown as faThumbsDownSolid,
 } from '@fortawesome/free-solid-svg-icons';
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { useAIState, useActions } from 'ai/rsc';
 
 import AnimatedButton from '@/components/ui/animatedButton';
+import { AI } from '@/lib/actions';
+import { GTagEvents, gtagEvent } from '@/lib/gtag';
 import { Feedback } from '@/models/chat';
 
+export type FeedbackLoadingStates = '' | Feedback;
+
 interface FeedbackButtonsProps {
-  feedback: Feedback | null;
-  onFeedback: (feedback: Feedback) => void;
-  disableFeedback: boolean;
+  onSharedPage: boolean;
 }
 
-const FeedbackButtons: React.FC<FeedbackButtonsProps> = ({
-  feedback,
-  onFeedback,
-  disableFeedback,
-}) => {
+const FeedbackButtons: React.FC<FeedbackButtonsProps> = ({ onSharedPage }) => {
+  const [aiState, setAIState] = useAIState<typeof AI>();
+  const { id: chatId, reaction: feedback } = aiState;
+  const { submitFeedback } = useActions<typeof AI>();
+
+  const onFeedback = async (newFeedback: Feedback) => {
+    if (onSharedPage) {
+      return;
+    }
+    // optimistically update the AI state
```

Comment:
i debated if this should wait until submitFeedback returns, but i figured it's nicer to have it look like the feedback immediately went through

Reply:
agreed. nicer and seems to be the common pattern in other apps.  but if it fails you should probably "unset" the feedback so they can try again